src/errors.rs

```diff
@@ -504,6 +505,37 @@ impl Display for S3ErrorCode {
         <Self as fmt::Debug>::fmt(self, f)
     }
 }
+
+/// Codegen for round trip between `S3ErrorCode` enum and string
+macro_rules! roundtrip_variant_and_str
+{
+    [$($v:tt,)+] => {
+
+        /// Returns a corresponding string of the error code
+        #[must_use]
+        pub const fn as_static_str(self) -> &'static str {
+            match self {
+                $(
+                    Self::$v => stringify!($v),
+                )+
+            }
+        }
+
+        /// Returns a corresponding error code from a string
+        #[must_use]
+        pub fn parse_from_str(s: &str) -> Self {
+            match s {
+                $(
+                   stringify!($v) =>  Self::$v,
+                )+
+                _ => {
+                    panic!("Unsupported error code: {}", s);
+                }
+            }
```

Comment:
It's better to return a result instead of panicking. You could add a simple error type like `ParseS3ErrorCodeError`.